File: backtracking/letter-case-permutation/letter-case-permutation.java
class Solution {
  public List<String> letterCasePermutation(String S) {
    List<String> res = new ArrayList<>();
    StringBuilder list = new StringBuilder();

    build(S, 0, list, res);

    return res;
  }

  private void build(String s, int i, StringBuilder list, List<String> res) {
    if (i == s.length()) {
      res.add(list.toString());
      return;
    }

    char curr = s.charAt(i);

    if (Character.isLetter(curr)) {
      list.append(Character.toLowerCase(curr));
      build(s, i + 1, list, res);
      list.deleteCharAt(list.length() - 1);

      list.append(Character.toUpperCase(curr));
      build(s, i + 1, list, res);

      list.deleteCharAt(list.length() - 1);
    } else {
      list.append(curr);
      build(s, i + 1, list, res);

      list.deleteCharAt(list.length() - 1);
    }
  }
}

File: backtracking/letter-case-permutation/letter-case-permutation.go
package lettercasepermutation

import "unicode"

func letterCasePermutation(S string) []string {
	res := []string{}
	list := []byte{}

	build(S, 0, &list, &res)

	return res
}

func build(s string, i int, list *[]byte, res *[]string) {
	if i == len(s) {
		*res = append(*res, string(*list))
		return
	}

	curr := s[i]

	if unicode.IsLetter(rune(curr)) {
		*list = append(*list, byte(unicode.ToLower(rune(curr))))
		build(s, i+1, list, res)
		*list = (*list)[:len(*list)-1]

		*list = append(*list, byte(unicode.ToUpper(rune(curr))))
		build(s, i+1, list, res)
		*list = (*list)[:len(*list)-1]
	} else {
		*list = append(*list, curr)
		build(s, i+1, list, res)
		*list = (*list)[:len(*list)-1]
	}
}
